dynamo/src/main/scala/d4s/codecs/D4SDecoder.scala

```diff
@@ -1,131 +1,55 @@
-            }.map(_.result())
-      }
-  }
-
-  implicit def mapLikeDecoder[V, M[_, _] <: Map[String, V]](
-    implicit
-    V: D4SAttributeDecoder[V],
-    factory: Factory[(String, V), M[String, V]]
-  ): D4SAttributeDecoder[M[String, V]] = {
-    attr =>
-      Either.fromTry(Try(attr.m())) match {
-        case Left(error) => Left(new CannotDecodeAttributeValue(s"Cannot decode $attr as Map", Some(error)))
-        case Right(value) =>
-          value.asScala.iterator
-            .foldLeft[Either[DynamoDecoderException, mutable.Builder[(String, V), M[String, V]]]](Right(factory.newBuilder)) {
-              case (acc, (name, attr)) =>
-                V.decodeAttribute(attr) match {
-                  case Left(error) => Left(error)
-                  case Right(value) =>
-                    acc.map(_ ++= Iterable((name -> value)))
-                }
-            }.map(_.result())
-      }
-  }
-
-  implicit def optionDecoder[T](implicit T: D4SAttributeDecoder[T]): D4SAttributeDecoder[Option[T]] = {
-    attr: AttributeValue =>
-      if (attr.nul()) Right(None) else T.decodeAttribute(attr).map(Some(_))
-  }
-}
 
 trait D4SDecoder[T] extends D4SAttributeDecoder[T] {
+  self =>
   def decode(item: Map[String, AttributeValue]): Either[DynamoDecoderException, T]
   def decode(item: java.util.Map[String, AttributeValue]): Either[DynamoDecoderException, T] = decode(item.asScala.toMap)
 
-  override final def decodeAttribute(item: AttributeValue): Either[DynamoDecoderException, T] = {
+  def decodeAttribute(item: AttributeValue): Either[DynamoDecoderException, T] = {
     Option(item.m())
       .toRight(new CannotDecodeAttributeValue(s"Couldn't decode dynamo item=$item as object. Does not have an `M` attribute (not a JSON object)", None))
       .flatMap(decode)
   }
+
+  def flatMap[T1](f: T => D4SDecoder[T1]): D4SDecoder[T1]                  = item => self.decode(item).map(f).flatMap(_.decode(item))
+  def map[T1](f: T => T1): D4SDecoder[T1]                                  = item => decode(item).map(f)
+  def map2[T1, A](another: D4SDecoder[T1])(f: (T, T1) => A): D4SDecoder[A] = flatMap(t => another.map(t1 => f(t, t1)))
 }
 
 object D4SDecoder {
   def apply[A: D4SDecoder]: D4SDecoder[A] = implicitly
+  def derived[T]: D4SDecoder[T] = macro CastedMagnolia.genWithCast[T, D4SDecoder[_]]
 
   def decode[A: D4SDecoder](item: Map[String, AttributeValue]): Either[DynamoDecoderException, A]           = D4SDecoder[A].decode(item)
   def decode[A: D4SDecoder](item: java.util.Map[String, AttributeValue]): Either[DynamoDecoderException, A] = D4SDecoder[A].decode(item)
   def decodeAttribute[A: D4SAttributeDecoder](v: AttributeValue): Either[DynamoDecoderException, A]         = D4SAttributeDecoder[A].decodeAttribute(v)
 
-  type Typeclass[T] = D4SAttributeDecoder[T]
-  def derived[T]: D4SDecoder[T] = macro CastedMagnolia.genWithCast[T, D4SDecoder[Nothing]]
+  implicit def fromCodec[T](implicit self: D4SCodec[T]): D4SDecoder[T] = self.decoder
```

Comment:
Better make D4SCodec a subtype again

Reply:
Yup, but in this case, D4SCodec will have all methods from encoder and decoder (combine/map2/...). Isn't it meaningless?

Reply:
Yup, but in this case, D4SCodec will have all methods from encoder and decoder (combine/map2/...). Isn't it meaningless?

Reply:
You can override them to return a more precise type `D4SCodec[A]` - this is legal ("covariant overrides") - then on the contrary it would be pretty convenient

Reply:
Hm, yup makes sense. Thanks !

Reply:
Hmm, actually you can't. You have to change the decoding part too to change the type parameter, so `map` can't return a `D4SCodec[A]`. Still I think it's acceptable...

Reply:
hm, nope
actually we cannot just rewrite those
all those methods are needed to change only one Decoder or Encoder
But Flatmap will work only for decoder (and contramap only for encoder), so we can not have codec with differrent types in encoder/decoder.

Reply:
Only in a case if we will preserve in codec both types or discard them. But it makes no sense .